test/studies/shootout/submitted/knucleotide4.chpl

```diff
@@ -0,0 +1,140 @@
+/* The Computer Language Benchmarks Game
+   https://salsa.debian.org/benchmarksgame-team/benchmarksgame/
+
+   contributed by Ben McDonald
+   derived from the Chapel #3 version by Ben Harshbarger and Brad Chamberlain
+*/
+
+use IO, Map, Sort;
+
+config param columns = 61;
+
+
+proc main(args: [] string) {
+  // Open stdin and a binary reader channel
+  const consoleIn = openfd(0),
+        fileLen = consoleIn.size,
+        stdinNoLock = consoleIn.reader(kind=ionative, locking=false);
+
+  // Read line-by-line until we see a line beginning with '>TH'
+  var buff: [1..columns] uint(8),
+      lineSize = 0,
+      numRead = 0;
+
+  while stdinNoLock.readline(buff, lineSize) && !startsWithThree(buff) do
+    numRead += lineSize;
+
+  // Read in the rest of the file
+  var dataDom = {1..fileLen-numRead},
+      data: [dataDom] uint(8),
+      idx = 1;
+
+  while stdinNoLock.readline(data, lineSize, idx) do
+    idx += lineSize - 1;
+
+  // Resize our array to the amount actually read
+  dataDom = {1..idx};
+
+  // Make everything uppercase
+  forall d in data do
+    d -= ("a".toByte() - "A".toByte());
+
+  writeFreqs(data, 1);
+  writeFreqs(data, 2);
+  writeCount(data, "GGT");
+  writeCount(data, "GGTA");
+  writeCount(data, "GGTATT");
+  writeCount(data, "GGTATTTTAATT");
+  writeCount(data, "GGTATTTTAATTTATAGT");
+}
+
+
+proc writeFreqs(data, param nclSize) {
+  const freqs = calculate(data, nclSize);
+
+  // create an array of (frequency, sequence) tuples
+  var arr = for (s,f) in freqs.items() do (f,s);
+
+  // print the array, sorted by decreasing frequency
+  for (f, s) in arr.sorted(reverseComparator) do
+   writef("%s %.3dr\n", decode(s, nclSize),
+           (100.0 * f) / (data.size - nclSize));
+  writeln();
+}
+
+
+proc writeCount(data, param str) {
+  const strBytes = str.bytes(),
+        freqs = calculate(data, str.numBytes),
+        d = hash(strBytes, strBytes.domain.low, str.numBytes);
+
+  writeln(freqs[d], "\t", decode(d, str.numBytes));
```

Comment:
Here, you could imagine doing either:

```chapel
        d = hash(strBytes, strBytes.domain.low, str.numBytes).val;
```

or:

```
  writeln(freqs[d], "\t", decode(d.val, str.numBytes));
```

(where I'd probably pick the former since it's more similar to what version 3 does).  But then I'm also wondering... was I wrong to make `hash` return a `hashVal` given that all we're going to do with it here is turn it back into an int?  Generally, I like this approach overall, but I'm struggling to remember what the code here is doing and why it doesn't use the hash to index into the map or anything...  Is it clearer to you?

Reply:
Hmm, I'm not sure I understand what you mean here. Is it not indexing into the map using the hash with the `freqs[d]` access?

I think that is why we wouldn't want to unwrap it during the `d = hash(strBytes, strBytes.domain.low, str.numBytes).val;` stage, because we need it as a `hashVal` for the `freqs[d]` access below.

Reply:
Oops, yes, you're right—I missed that `d` was used in two places.